### PR TITLE
[Server] Separate the explicit MeshSync deployment mode from its cache

### PR DESCRIPTION
**Pull request by CodeAhmedJamil**
## Two defects, one root cause

`connection.metadata.meshsync_deployment_mode` was carrying **two different
facts** under one key:

1. the user's explicit per-connection choice, written by the Connection Wizard's
   MeshSync Mode step and the kubeconfig import picker, and
2. a materialized cache of the resolved mode, rewritten by the controllers editor
   on every save, which every pre-layered consumer reads (the connection state
   machine, the header status chips, the kubeconfig flows).

Meshery writes that key on **every** Kubernetes connection at registration. So
treating it as fact (1) made it outrank every other layer on every connection,
and two visible defects follow directly:

- **A server-wide deployment mode could never reach an existing connection.**
  `applyControllersConfigToConnection` resolved the mode from that metadata key
  first, and the key is always present, so the fan-out from
  `UpdateControllersDefaultConfig` silently kept the old mode. The Settings
  control appeared to work and changed nothing.
- **The wizard and the controllers editor disagreed.** The wizard wrote the
  legacy key alone and left `controllers_config.operator.deploymentMode`
  untouched, so the editor afterwards showed an "Override" chip for a mode the
  connection was not running.

## The fix, at the layer the defect lives on

`server/models/connections/deployment_mode_resolution.go` separates the two
meanings and owns the single definition of "which mode is this connection
running, and why":

- `controllers_config.operator.deploymentMode` is now the **only** store of the
  explicit choice. Every entry point writes it through
  `SetDeploymentModeOverride`, so the wizard and the editor cannot disagree.
- `meshsync_deployment_mode` is now **only** the materialized cache, written
  through `MaterializeMeshsyncDeploymentMode`, and it participates in resolution
  strictly *below* the layered document - as a compatibility floor so connections
  registered before the layered document existed keep their recorded mode. It can
  no longer mask a server-wide default.

`ResolveDeploymentMode` also reports **which layer** supplied the mode, so the UI
can say where an effective mode came from instead of presenting it as an
unexplained fact.

## Why these are one PR and not two

They are one root cause wearing two faces, and splitting them produces a strictly
worse intermediate state than shipping neither: fixing the precedence alone,
while the wizard still writes the legacy key directly, would make the editor and
the wizard disagree **more** than they do today.

## Tests

`deployment_mode_resolution_test.go`, 14 subtests, named for the defects they
pin - including the edge cases: a malformed stored override surfaces as an error
rather than a silent overwrite, clearing the mode prunes an empty override
document rather than leaving one behind, and writing the mode preserves sibling
fields such as `operator.version`.

`go build ./...` exit 0; `go test ./server/...` green.

## Dependencies

**Depends on #21144** (the settings-surface map), because its documentation commit
edits the page that PR adds. Independent of #21145.

## Summary by CodeRabbit

- **New Features**
  - Added layered configuration for Operator, MeshSync, and Broker settings.
  - Supports built-in defaults, server-wide settings, and per-connection overrides with clear precedence.
  - Configuration changes propagate to connected Kubernetes environments, including necessary MeshSync refreshes.
  - Operator chart versions can be customized and reconciled per connection.
  - Settings screens display effective deployment modes, applicable fields, dormant settings, and discard options.

- **Bug Fixes**
  - Improved validation, error reporting, inheritance handling, and synchronization with server-resolved values.
  - Updates are applied atomically and safely handle invalid or unavailable resources.

- **Documentation**
  - Added guidance covering configuration behavior, propagation, validation, and known limitations.

### CODE REVIEW

**Review by pontusringblom — COMMENTED** [review]
Does this need a change in the MeshSync CRD, too?

File: docs/content/en/project/contributing/contributing-controllers-config.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+- **`controllers_config.operator.deploymentMode`** is the only store of the
+  explicit choice. Every entry point - the wizard's MeshSync Mode step, the
+  kubeconfig import picker, the controllers editor - writes it through
+  `connections.SetDeploymentModeOverride`, so no two controls can disagree
+  about what a connection is set to.
+- **`meshsync_deployment_mode`** is only the materialization of the *resolved*
+  mode, written through `connections.MaterializeMeshsyncDeploymentMode` and
+  read by the consumers that predate the layered document (the connection state
+  machine, the header status chips, the kubeconfig flows).
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply docs/content/en/project/contributing/contributing-controllers-config.md:231]: Accepted - the wizard and the connections table both read metadata.meshsyncDeploymentMode ?? metadata.meshsync_deployment_mode, so both spellings are accepted on read while the server writes the snake_case form. Documenting both and stating explicitly that they are one materialized value, not two settings.

File: server/handlers/controllers_config_handlers.go [hunk 2]
```diff
@@ -398,11 +393,82 @@ func (h *Handler) reapplyControllersConfigToTrackedConnections(token string, use
...
+		h.log.Error(err)
+		override = nil
+	}
+	merged, _ := connections.ResolveControllersConfig(override, serverDefaults)
+	desired := connections.ResolveDeploymentMode(merged, metadata, h.MeshsyncDefaultDeploymentMode)
+
+	previous := connections.MeshsyncDeploymentModeFromMetadata(metadata)
+	if previous == desired.Mode {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply server/handlers/controllers_config_handlers.go:470]: Accepted, and this is the most serious item raised. A connection whose stored override merely fails to parse would resolve to the server-wide default, differ from the previous materialization, and get reconciled - tearing down and redeploying MeshSync into a mode the user never chose, on the strength of corrupt data. That is a silent destructive action taken on an unreadable input, which is exactly the failure class this PR exists to remove. An unparseable override is unknown intent, not absent intent: reconciliation should be skipped for that connection and the parse failure surfaced, which the per-connection GET already does.

File: ui/components/configuration/__tests__/ControllersConfigForm.test.tsx [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+      ['Replicas', brokerGrid()],
+      ['Service type', undefined],
+      ['External endpoint override', undefined],
+      ['Service annotations', undefined],
+    ];
+    for (const [label, scope] of inertFields) {
+      expect(disabledState(control(label, scope)), `${label} should be inert`).toBe(true);
+      expect(field(label, scope)).toHaveTextContent('Not applied');
+    }
+  });
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply ui/components/configuration/__tests__/ControllersConfigForm.test.tsx:435]: Confirmed and resolved, though the cause was not where it looked. `operator.version` is correctly absent from `EMBEDDED_LIVE_PATHS`, so `takesEffectIn('embedded', ['operator','version'])` is false and the form does render it inert - the mode table is right and the test's expectation is right. The failure was mine: while distributing review fixes across the stacked branches I replayed test and component blobs onto this branch without the commits they depend on, leaving it with `fieldPath.ts` but no `useControllersConfigDraft.ts` and the deleted RTK layer still present. The form could not behave as the test expected because the form was a mixture of two states. Fixed by merging #21145 into this branch, which is the dependency that should always have been there: B1's mode-governed form builds directly on the simplified editor. `npx vitest run components/configuration` now passes 31/31 and `go build ./...` is clean.

File: ui/tests/e2e/controllers-config.spec.ts [hunk 4]
```diff
@@ -0,0 +1,816 @@
...
+    // `page` is required so Playwright builds the browser context for cases that
+    // only use the API fixtures too - keeping the reset identical across cases.
+    expect(page).toBeTruthy();
+  });
+
+  test.afterAll(async ({ request }) => {
+    // Leave the server as the run found it: the defaults are server-wide state
+    // that outlives this spec and fans out to every tracked connection.
+    await clearServerDefaults(request);
+  });
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply ui/tests/e2e/controllers-config.spec.ts]: Accepted - and worse than cosmetic. `test.afterAll` receives only worker-scoped fixtures, so taking the test-scoped `request` context made the cleanup hook throw, which means the server-wide defaults this spec sets were being left behind for whatever ran next. That is precisely the cross-test contamination the hook exists to prevent, and it would have been a confusing failure to attribute. It now builds its own context with `playwright.request.newContext({ baseURL })` and disposes it in a `finally`. Commit a4541348802d.

File: docs/content/en/project/contributing/contributing-controllers-config.md [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+  single paste for that reason; fixing the controls means holding the raw text
+  in component state and parsing out of it.
+
+Still missing, and required before this surface can be called covered:
+
+- There is no end-to-end spec. When one is added it belongs at
+  `ui/tests/e2e/controllers-config.spec.ts`, keyed to the **Operator, MeshSync &
+  Broker Settings** Test Plan Test Group via the Allure `testGroup` label - the
+  same mechanism the Connection Lifecycle report uses (see
+  [Contributing to Meshery's CLI tests]({{< relref "cli/tests.md" >}})).
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply docs/content/en/project/contributing/contributing-controllers-config.md]: Accepted - the section was stale the moment this PR added the spec. It now describes what exists (the end-to-end spec keyed to the Operator, MeshSync & Broker Settings Test Group, the rendered-form vitest, and the cluster-propagation test that `controllers_config_apply.go` previously lacked entirely) and lists only what genuinely remains uncovered: the live round trip against a real operator, which the spec self-skips when no cluster is reachable rather than passing vacuously. Commit a4541348802d.

File: ui/components/configuration/ControllersConfigForm.tsx [hunk 6]
```diff
@@ -13,7 +13,26 @@ import {
...
+/**
+ * The editable controllers configuration document: the generated PUT request
+ * body, i.e. the wire document without the server-stamped `schemaVersion`.
+ * Both layers (server-wide defaults and the per-connection override) accept
+ * this exact shape.
+ */
+export type ControllersConfigDoc = UpdateControllersDefaultConfigApiArg['body'];
+
+/** As returned by the server, which stamps the schema revision onto the document. */
+export type VersionedControllersConfigDoc = ControllersConfigDoc & { schemaVersion?: string };
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply ui/components/configuration/ControllersConfigForm.tsx:35]: Declined - false positive, and I checked rather than assuming. There is exactly one declaration: `git show <branch>:ui/components/configuration/ControllersConfigForm.tsx | grep -c '^export type VersionedControllersConfigDoc'` returns 1, and `npx tsc --noEmit` against the branch content exits 0 with zero TS2300 duplicate-identifier errors. I suspect this was analysed against the transient merge-conflicted state of the file, which did briefly contain both sides while I merged #21145 in to make the branch self-consistent; the committed state never did.

File: server/models/meshery_controllers.go [hunk 7]
```diff
@@ -85,6 +85,13 @@ type MesheryControllersHelper struct {
 	// configuration changes; consumed by the embedded meshsync run options.
 	controllersConfig *controllersconfig.MesheryControllersConfig
 
+	// attachedOperatorChartVersion is the Meshery Operator Helm chart version the
+	// currently attached operator controller handler was constructed with. The
+	// meshkit handler captures its deployment config at construction, so this is
+	// the only record of what a later `operator.version` change has to differ
+	// from before a redeploy is warranted. "" until handlers are attached.
+	attachedOperatorChartVersion string
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply server/models/meshery_controllers.go:93]: Accepted - real, and a nasty failure mode. AddCtxControllerHandlers sets attachedOperatorChartVersion = depConfig.MesheryReleaseVersion at attach time, and ReconcileOperatorChartVersion calls it before Deploy. So a Helm failure left the desired version already recorded as attached, and the equality guard at the top of the function then short-circuited every later call - the failed upgrade became the permanent resting state with no retry path, which is worse than the upgrade never having been attempted. The previous value is now restored on all three failure paths (setup error, handler not attached, deploy error). Commit 2a28eb2a0022.

File: server/handlers/connections_handlers.go [hunk 8]
```diff
@@ -726,11 +726,25 @@ func (h *Handler) handleMeshSyncDeploymentModeChange(
...
+			} else {
+				ctrlHelper.SetControllersConfig(merged)
+			}
 			ctrlHelper.
-				AddCtxControllerHandlers(machineCtx.K8sContext).
 				SetMeshsyncDeploymentMode(newMeshSyncMode).
+				AddCtxControllerHandlers(machineCtx.K8sContext).
 				UpdateOperatorsStatusMap(machineCtx.OperatorTracker).
 				DeployUndeployedOperators(machineCtx.OperatorTracker).
 				AddMeshsyncDataHandlers(ctx, machineCtx.K8sContext, userID, mesheryInstanceID, provider)
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply server/handlers/connections_handlers.go:754]: Accepted. Resolution ran after the operators were undeployed, and on failure the code logged and reattached from the last-known-good document - which is a partial reconciliation performed on unreadable input. That is the same fault as reconciling a connection whose override will not parse, which this PR fixes a few files away, so declining here would have been inconsistent with the principle the PR is built on: do not take a destructive action on data you could not read. Resolution now happens before the teardown block and returns the error with the connection left exactly as it was. Commit 2a28eb2a0022.

File: ui/components/configuration/deploymentMode.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+  // mode is whatever Meshery Server resolves on save, and saying so is more
+  // honest than guessing at a floor the client cannot see.
+  return {
+    mode: effectiveMode,
+    sourceLabel: unsaved
+      ? 'the Meshery Server default, resolved when you save'
+      : 'the Meshery Server default',
+    unsaved,
+    scope: 'connection',
+  };
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply ui/components/configuration/deploymentMode.ts:189]: Acknowledged as a real edge case, declined for this PR as scope expansion - and you graded it Heavy lift yourself, which I think is the right read. You are correct that after a user clears an override, serverEffective still reflects the pre-clear state, so the form can briefly gate on a mode the server will not resolve to. The code already refuses to pretend otherwise: connectionDeploymentMode returns sourceLabel 'the Meshery Server default, resolved when you save' whenever the change is unsaved, precisely because the client cannot see the MESHSYNC_DEFAULT_DEPLOYMENT_MODE floor beneath the document. Representing this as a distinct unresolved state, or returning the post-clear resolution from the API, is a genuine improvement but it changes the API contract and the form's state model, which does not belong in a PR whose subject is separating the deployment mode's two stores. Tracking it rather than doing it here.

File: docs/content/en/project/contributing/contributing-controllers-config.md [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+- `ui/components/configuration/__tests__/deploymentMode.test.ts` - which
+  settings each deployment mode can apply, and how the mode governing each
+  editor is resolved and attributed to a layer.
+
+Still missing, and required before this surface can be called covered:
+
+- `ui/components/configuration/__tests__/ControllersConfigForm.test.tsx` - the
+  rendered editor: tri-state inherit/override, the conditional LoadBalancer-only
+  fields, and the deployment-mode gating (inert-and-explained on a connection,
+  annotated-but-live on the server-wide defaults).
```
[the remark this answers is not in the document]
Comment (CodeAhmedJamil) [reply docs/content/en/project/contributing/contributing-controllers-config.md:303]: Accepted, and thank you for catching the half-fix. My previous commit replaced the bullets under 'Still missing, and required before this surface can be called covered' but left the heading in place, so the page listed the rendered-form, cluster-application and handler tests - all three added by this PR - as still absent. Strictly worse than the original stale text, since it now contradicted the diff directly. They are part of the unit coverage list now, and only the live-cluster round trip remains under what is uncovered.

**Review by hortison — APPROVED** [review]
Reviewed as a maintainer.

- Change: 31 file(s), +5177/-158, base master
- Checks at review time: 1 skipped, 20 success

Reviewed the diff and the check results. Approving.
